packages/api/internal/dns/server.go

```diff
@@ -2,99 +2,242 @@ package dns
 
+const cachedDnsPrefix = "sandbox.dns."
+
 type DNS struct {
-	mu      sync.Mutex
-	redis   *redis.Client
-	records *smap.Map[string]
+	srv    *resolver.Server
+	logger *zap.SugaredLogger
+
+	remote *cache.Cache
+	local  *smap.Map[string]
+
+	closer struct {
+		once sync.Once
+		op   func(context.Context) error
+		err  error
+	}
 }
 
-func New(rc *redis.Client) *DNS {
-	return &DNS{
-		redis:   rc,
-		records: smap.New[string](),
+func New(ctx context.Context, rc *redis.Client, logger *zap.SugaredLogger) *DNS {
+	d := &DNS{logger: logger}
+
+	if rc != nil {
+		d.remote = cache.New(&cache.Options{Redis: rc, LocalCache: cache.NewTinyLFU(10_000, time.Hour)})
+	} else {
+		d.local = smap.New[string]()
 	}
+
+	return d
 }
 
-func (d *DNS) Add(_ context.Context, sandboxID, ip string) {
-	d.records.Insert(d.hostname(sandboxID), ip)
+func (d *DNS) Add(ctx context.Context, sandboxID, ip string) {
+	switch {
+	case d.remote != nil:
+		d.remote.Set(&cache.Item{
+			Ctx:   ctx,
+			TTL:   redisTTL,
+			Key:   sandboxID,
+			Value: ip,
+		})
+	case d.local != nil:
+		d.local.Insert(sandboxID, ip)
+	}
 }
 
-func (d *DNS) Remove(_ context.Context, sandboxID, ip string) {
-	d.records.RemoveCb(d.hostname(sandboxID), func(key string, v string, exists bool) bool {
-		return v == ip
-	})
+func (d *DNS) Remove(ctx context.Context, sandboxID, ip string) {
+	switch {
+	case d.remote != nil:
+		if err := d.remote.Delete(ctx, d.cacheKey(sandboxID)); err != nil {
+			d.logger.Debug("removing item from DNS cache", zap.Error(err), zap.String("sandbox", sandboxID))
+		}
+	case d.local != nil:
+		d.local.RemoveCb(d.cacheKey(sandboxID), func(k string, v string, ok bool) bool { return v == ip })
+	}
 }
 
-func (d *DNS) get(hostname string) (string, bool) {
-	return d.records.Get(hostname)
+func (d *DNS) Get(ctx context.Context, sandboxID string) net.IP {
+	var res string
+	switch {
+	case d.remote != nil:
+		if err := d.remote.Get(ctx, d.cacheKey(sandboxID), &res); err != nil {
+			if errors.Is(err, cache.ErrCacheMiss) {
+				d.logger.Warn("item missing in remote DNS cache", zap.String("sandbox", sandboxID))
+			} else {
+				d.logger.Error("resolving item from remote DNS cache", zap.String("sandbox", sandboxID), zap.Error(err))
+			}
+		}
+	case d.local != nil:
+		var ok bool
+		res, ok = d.local.Get(d.cacheKey(sandboxID))
+		if !ok {
+			d.logger.Warn("item not found in local DNS cache", zap.String("sandbox", sandboxID))
+		}
+	}
+
+	addr := net.ParseIP(res)
+	if addr == nil {
+		if res != "" {
+			d.logger.Error("malformed address in cache", zap.Bool("local", d.local != nil), zap.String("addr", res))
+		}
+
+		addr = net.ParseIP(defaultRoutingIP)
+	}
+
+	return addr.To4()
 }
 
-func (*DNS) hostname(sandboxID string) string {
-	return fmt.Sprintf("%s.", sandboxID)
+func (d *DNS) cacheKey(id string) string {
+	switch {
+	case d.remote != nil:
+		return fmt.Sprintf("%s%s", cachedDnsPrefix, id)
+	case d.local != nil:
+		return fmt.Sprintf("%s.", id)
```

Comment:
adding the terminal dot for the cache keys is a thing that happened previously, and I've retained that: the "hostname" value isn't used outside of the cache, so I'm not sure what benefit we get from this, but if it's cruft we should remove.

Reply:
I think it's based on `.` used in `dig`

Reply:
Okay this should explain it
<img width="856" alt="image" src="https://github.com/user-attachments/assets/20baeeab-7dd6-45c6-af3c-c45655914466" />

Reply:
Looking more into it, I would strip it in `handleDNSRequest` and then remove it everywhere, it's just confusing

Reply:
I think the errors you saw were because I wasn't putting the cacheKey on the on when inserting into the remote cache.